### ISSUE TITLE
"keepIncorrectValues": true is ignored

### ISSUE THREAD

**Issue by masciugo** [surveyjs/survey-library#6489]
### Are you requesting a feature, reporting a bug or asking a question?

bug

### What is the current behavior?

when I use `"keepIncorrectValues": true` for a dropdown question and I run `survey.clearIncorrectValues();` the value (that is not present in the choices array) is cleared

https://plnkr.co/edit/cHNJ3CqMAO5p7ydx

### What is the expected behavior?

I would expect not as I am saying to keep it

**Comment by andrewtelnov** [comment]
@masciugo `survey.keepIncorrectValues` property is not serializable.  You have to set it in the code:
``survey.keepIncorrectValues = true;`` The reason we make it non-serializable, because a developer should decide should a survey/form keeps incorrect values or not. A Creator user should not think about it.

Thank you,
Andrew

**Comment by masciugo** [comment]
@andrewtelnov thanks for the explanation. I can run programmatically `question.keepIncorrectValues = true` but the problem is reaching every questions in the survey to do that. Some of them are really nested and may not be rendered a soon as the survey is rendered  WITH PRELOADED DATA because belongs to collapsed panel or other pages, they are not visible. I wrote a little method to collect all questions: 

```
  allAtomicQuestions = (survey) => {
    let allAtomicQuestions = survey.getAllQuestions()
    for (let q of allAtomicQuestions) {
      switch(q.getType()) {
      case "paneldynamic":
        for (let item of q.items) {
          Array.prototype.push.apply(allAtomicQuestions, item.panel.elements)
        }
        break
      case "matrixdynamic": {
        for (let row of q.generatedVisibleRows || []) {
          for (let cell of row.cells) {
            allAtomicQuestions.push(cell.question)
          }
        }
        break
      }
      default:
      }
    }
    return allAtomicQuestions
  }
```

It would be but the matrixdynamic part which only works if the cell questions are rendered. Basically I cannot do what I can do with paneldynamic items. Do you know how can do that? Which matrixdynamic method would return the cell questions of all rows?

thank you for your help

**Comment by andrewtelnov** [comment]
@masciugo You can set it on survey level, as I wrote above:
`survey.keepIncorrectValues = true;`

Thank you,
Andrew

**Comment by masciugo** [comment]
thank you Andrew but I need to set keepIncorrectValues to true only for some questions, the one that I created as custom widgets.  I could do something really similar to what you do [here](https://github.com/surveyjs/custom-widgets/blob/master/src/select2-tagbox.js#L72-L74). The problem is that the customisation takes place at rendering time and my custom widget questions are not rendered at the time I want to tell it to keepIncorrectValues. That's why I wrote allAtomicQuestions method. 

basically I'd like in order to:

1. render the survey with data
2. tell some questions (including the not visible) to keep incorrect values
3. survey.clearIncorrectValues()
4. survey.validate() and show errors to the users

So the only problem is that I am not able to obtain cell questions for matrixdynamics as I am able to with paneldynamics

thanks

**Comment by andrewtelnov** [comment]
@masciugo You can make this property serializable:
`Survey.Serializer.addProperty("selectbase", {name:"keepIncorrectValues:boolean", category: "data"});`

Thank you,
Andrew

**Comment by masciugo** [comment]
ok it works but fact is that my app users, which create their survey, should not deal with this kind of detail (as you already told me). Is there a way to say in general: _if a question has some property than set another property to something_?

I think that would solve my problem

exploiting something like `question.registerPropertyChangedHandlers` but, again, I would reach the question...

thanks again

**Comment by andrewtelnov** [comment]
@masciugo Sure question.registerPropertyChangedHandlers will work. You can use survey.onQuestionCreated event to update the property or set the handler. However the survey model should be created differently:
```js
// Create an empty model
const survey = new Survey,Model();
survey.onQuestionCreated.add((sender, options) => { /*Code*/});
// Load the survey JSON schema
survey.fromJSON(surveyJson);
```

Thank you,
Andrew

**Comment by masciugo** [comment]
hi @andrewtelnov I tried with your suggestion which sounds good and clean

https://plnkr.co/edit/611xHEqd3GePEkfY

But it seems that `myCustomProp` is not correctly set up with the value specified in the model (true) when the question is created

thank you

**Comment by andrewtelnov** [comment]
@masciugo I mean by adding this code:
```jsSurvey.Serializer.addProperty("selectbase", {
    name: "keepIncorrectValues",
     type: "boolean"
});
```
Here is [your example](https://plnkr.co/edit/fqQzioSZj1Zue67T).

Thank you,
Andrew

**Comment by masciugo** [comment]
[Yes I had already tried this suggestion and it works](https://github.com/surveyjs/survey-library/issues/6489#issuecomment-1635997542). But I would like to set keepIncorrectValues programmatically and, above all, conditionally. The condition is based on another question prop (called myCustomProp) and the logic is as simple as: if myCustomProp is true then set keepIncorrectValues to true. (this is a simplification)

So you suggested me to run my logic inside the `onQuestionCreated` hook. The problem is that when the question is created and the code is run the `question.myCustomProp` is set to its default value (false) and not the one I chose in the json model (true)! you can see that at line 14 of my plunk 

It seems that `question.myCustomProp` is set to true someway later than the `onQuestionCreated` event

why?

thank you

**Comment by andrewtelnov** [comment]
@masciugo I remember I wanted to change it to speed up the loading from JSON. Load an object first and then added it into a parent array. However, hundreds unit tests turned red and I was not able to fix them. As result I improved the performance by introducing `isLoadingFromJSON` property into base object and it works even better, since it allows to kill all unneeded notifications, including adding an object into a parent array.
Anyway, here is another suggestion. You can create a wrapper proeprty by using "onGetValue/onSetValue" callback functions in property defintion:
```js
Survey.Serializer.addProperty("selectbase", {
    name: "myCustomProp",
     type: "boolean",
      default: false,
      onGetValue: function (obj) {
        return obj.keepIncorrectValues;
      },
      onSetValue: function (obj, value, jsonConv) {
        obj.keepIncorrectValues = value;
      }
});
```
Here is [the example](https://plnkr.co/edit/Axg3nDEHpifNpgEN).

Thank you,
Andrew

**Comment by masciugo** [comment]
Hi Andrew I think I solved with your last suggestion. Actually I want to control `keepIncorrectValues` prop value:

```
    {
      name: "keepIncorrectValues",
      type: "boolean",
      default: false,
      onGetValue: function (obj) {
        return obj.myCustomProp || obj.getPropertyValue("keepIncorrectValues");
      }
    },
```

About the fallback to the orginal user specified value (`obj.getPropertyValue("keepIncorrectValues")`) I tried it empirically and it works. But you can confirm that using `getPropertyValue` is the correct way to do that because bypass `onGetValue` callback generating an infinite loop?

**Comment by andrewtelnov** [comment]
@masciugo Yep, you will go infinitive loop, because `obj.getPropertyValue("keepIncorrectValues")` shold call `onGetValue` callback function.

Thank you,
Andrew
